### ISSUE TITLE
Ability to configure the reference frame

### ISSUE THREAD

**Issue by shotamatsuda** [takram-design-engineering/three-geospatial#11]
Meter-scale ECEF coordinates in world space often suffer from floating-point precision issues unless compensated for separately. In a simple setup, it’s easier to work near the world space origin by transforming the reference frame.

**Comment by gkjohnson** [comment]
I think this is the same concept - but just to document the issue I ran in to that this sounds like it will address:

For my usecase I often transform the globe data such that the lat / lon surface point is located at the origin and oriented so the ellipsoid normal is aligned with three.js' y-up. The atmosphere shader doesn't seem to have a way to set the orientation / translation easily. I see `ellipsoid_center` but not `ellipsoid_rotation`. Being able to assign an ellipsoid matrix to adjust rotation and translation would be easier, I think.

> Meter-scale ECEF coordinates in world space often suffer from floating-point precision issues unless compensated for separately.

In the past I've overcome these kinds of precision issues by pre-transforming any data into the local camera frame before assigning to the uniforms so all the calculations are done in a space less likely to suffer from precision loss.

For example, I was seeing issues when working on a BatchedMesh renderer, for example, for globe tiles and saw that the transform-scale per mesh was causing the jittering issues. I created a [BatchedMesh extension class](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/master/example/src/plugins/batched/BatchedTilesPlugin.js) that transform all the BatchdMesh instances into the local camera frame before render to prevent that.

**Comment by shotamatsuda** [comment]
Yes, that’s exactly what I was thinking: adding a rotation matrix for the ellipsoid and making the center configurable.

I couldn’t figure out where the transformation happens before rendering in your plugin, so I’ll need to read through later. What I’m more interested in, though, is `ExpandingBatchedMesh` because that’s what I really need to implement. 😄

**Comment by gkjohnson** [comment]
> I couldn’t figure out where the transformation happens before rendering in your plugin

Oh sorry I thought linked to [this ModelViewBatchedMesh](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/a99b79469bcf09e46a1f79499bd86fb874fee989/example/src/plugins/batched/ModelViewBatchedMesh.js#L60-L80) class which replaces the regular mesh matrix texture and object matrixWorld with ones transformed into the local camera frame just before render. They are then replaced immediately after render.

> What I’m more interested in, though, is ExpandingBatchedMesh because that’s what I really need to implement. 😄

Happy to answer any questions about either!

**Comment by gkjohnson** [comment]
I'm thinking I may be interested in this sooner rather than later since working in oriented frame at the planet surface is less convenient than working at the origin frame.

Do you have any preference for how something like this might be added?

**Comment by shotamatsuda** [comment]
Sorry for my slow progress on this. I’ve been caught up with [rendering clouds](https://github.com/takram-design-engineering/three-geospatial/tree/dev/clouds/packages/clouds) lately.

The same correction is applied to clouds, but since I’m using temporal supersampling and jittering in projection / reprojection, offsetting the ellipsoid center has become complicated, let alone introducing another rotation matrix in the shaders. Instead, I’m considering adjusting the view matrix used in post-processing, although I don’t have a concrete idea yet. Perhaps it doesn’t work and I’ll just have to bare with the complication.

**Comment by shotamatsuda** [comment]
Started working on this in [feature/ellipsoid-center-matrix](https://github.com/takram-design-engineering/three-geospatial/tree/feature/ellipsoid-center-matrix). The coordinate conversion is almost done.

This update will add `ellipsoidMatrix: Matrix4` and exposes `ellipsoidCenter: Vector3` in both `AerialPerspectiveEffect` and `AtmosphereMaterialBase`. To move the Earth based on `longitude`, `latitude`, and `height`, the following setup will work:

```ts
// Given:
// const camera: Camera
// const aerialPerspective: AerialPerspectiveEffect

const geodetic = new Geodetic(radians(longitude), radians(latitude), height)
const position = geodetic.toECEF()
aerialPerspective.ellipsoidCenter.copy(position).multiplyScalar(-1)
const up = Ellipsoid.WGS84.getSurfaceNormal(position)
rotation.setFromUnitVectors(up, camera.up)
aerialPerspective.ellipsoidMatrix.makeRotationFromQuaternion(rotation)
```

I’m currently thinking about the R3F interface.

**Comment by shotamatsuda** [comment]
Added in https://github.com/takram-design-engineering/three-geospatial/commit/ee7ceae7af9efdf4dba097a90a3feb8482b7e2b0
You can see a working example at: https://takram-design-engineering.github.io/three-geospatial/?path=/story/atmosphere-atmosphere--moving-ellipsoid

**Comment by gkjohnson** [comment]
Oh amazing! Thank you! To be clear `ellipsoidMatrix` is _just_ for rotations? Rather than the position and rotation (and scale?) being encoded in the "ellipsoidMatrix" uniform?

> Sorry for my slow progress on this. I’ve been caught up with [rendering clouds](https://github.com/takram-design-engineering/three-geospatial/tree/dev/clouds/packages/clouds) lately.

The cloud work has been looking amazing - looking forward to it!

**Comment by shotamatsuda** [comment]
It’s treated as mat3 in the shaders, so translation is ignored. Scaling might work, but hasn’t been tested and may break some assumptions I made. I’ve made some attempts to merge the center and rotation into a matrix, but none have been successful so far. Will revisit it in the future.

**Comment by shotamatsuda** [comment]
Maybe it’s too early to close this. I was thinking adding another inertial reference frame.

**Comment by gkjohnson** [comment]
I just gave this a try and after trying to apply the same transform as the 3d-tiles-renderer group I landed on the following:

```js
const matrix = new Matrix4();

// ...

matrix.copy( tiles.group.matrixWorld ).setPosition( 0, 0, 0 );
atmosphere.ellipsoidMatrix.copy( matrix );
atmosphere.ellipsoidCenter.setFromMatrixPosition( tiles.group.matrixWorld ).applyMatrix4( matrix.invert() );
```

Specifically it looks like `ellipsoidCenter` needs to have the inverse of the "ellipsoidMatrix" rotation applied. Otherwise it works great!

For reference I wrapped the "Atmosphere" component [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/bfdac1ff7f654d2b3fad43b6c86699d623eccd13/example/r3f/atmosphere.jsx#L114-L123) to automatically use the tiles group transform. It should be possible to wrap or update the Atmosphere component to automatically use the parent transformation to drive the atmosphere transformations for convenience, as well.

**Comment by gkjohnson** [comment]
I just gave this a try and after trying to apply the same transform as the 3d-tiles-renderer group I landed on the following:

```js
const matrix = new Matrix4();

// ...

matrix.copy( tiles.group.matrixWorld ).setPosition( 0, 0, 0 );
atmosphere.ellipsoidMatrix.copy( matrix );
atmosphere.ellipsoidCenter.setFromMatrixPosition( tiles.group.matrixWorld ).applyMatrix4( matrix.invert() );
```

Specifically it looks like `ellipsoidCenter` needs to have the inverse of the "ellipsoidMatrix" rotation applied. Otherwise it works great!

For reference I wrapped the "Atmosphere" component [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/bfdac1ff7f654d2b3fad43b6c86699d623eccd13/example/r3f/atmosphere.jsx#L114-L123) to automatically use the tiles group transform. It should be possible to wrap or update the Atmosphere component to automatically use the parent transformation to drive the atmosphere transformations for convenience, as well.

**Comment by shotamatsuda** [comment]
Replaced `ellipsoidCenter` and `ellipsoidMatrix` by `worldToECEFMatrix` in https://github.com/takram-design-engineering/three-geospatial/commit/d841ef0c512c3f9f0081a5904740320279991cd3. They had to be breaking changes, but it's much better now.

```ts
geodetic.set(radians(longitude), radians(latitude), height)
Ellipsoid.WGS84.getNorthUpEastFrame(
  geodetic.toECEF(position),
  atmosphere.worldToECEFMatrix
)
```
